prov/rxm/src/rxm_conn.c

```diff
@@ -143,44 +159,61 @@ static void rxm_conn_close(struct util_cmap_handle *handle)
 	FI_DBG(&rxm_prov, FI_LOG_EP_CTRL,
 	       "Saved MSG EP fid for further deletion in main thread\n");
 	rxm_conn->msg_ep = NULL;
+	/* Now we ready to create new MSG EP */
 }
 
 static void rxm_conn_free(struct util_cmap_handle *handle)
 {
 	struct rxm_conn *rxm_conn = container_of(handle, struct rxm_conn, handle);
-
-	/* This handles case when saved_msg_ep wasn't closed */
-	if (rxm_conn->saved_msg_ep) {
-		if (fi_close(&rxm_conn->saved_msg_ep->fid))
-			FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
-				"Unable to close saved msg_ep\n");
+	struct rxm_ep *rxm_ep = container_of(handle->cmap->ep, struct rxm_ep, util_ep);
+	struct util_cmap_cmd *cmd;
+	struct rxm_cmap_cmd_data *cmd_data;
+
+	if (!rxm_ep->util_ep.cmap->event_handler_closing) {
+		cmd = calloc(1, sizeof(struct util_cmap_cmd) +
+				sizeof(struct rxm_cmap_cmd_data));
+		if (!cmd)
+			goto conn_free;
+		cmd->type = UTIL_CMAP_CMD_CONN_FREE;
+		cmd_data = (struct rxm_cmap_cmd_data *)cmd->data;
+		cmd_data->rxm_conn = rxm_conn;
+
+		dlist_ts_insert_tail(&rxm_ep->util_ep.cmap->cmd_queue, &cmd->entry);
+		rxm_ep->util_ep.cmap->cmd_write++;
+		rxm_conn_wake_up_wait_obj(rxm_ep);
+	} else {
+conn_free:
+		rxm_conn_close_msg_ep(rxm_ep, rxm_conn);
+		rxm_send_queue_close(&rxm_conn->send_queue);
+		free(rxm_conn);
 	}
+}
 
-	if (!rxm_conn->msg_ep)
+static void
+rxm_conn_process_connected(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn)
+{
+	struct util_cmap_cmd *cmd;
+	struct rxm_cmap_cmd_data *cmd_data;
+
+	cmd = calloc(1, sizeof(struct util_cmap_cmd) +
+			sizeof(struct rxm_cmap_cmd_data));
+	if (!cmd) {
+		if (!rxm_ep->srx_ctx) {
+			(void) rxm_ep_prepost_buf(rxm_ep, rxm_conn->msg_ep);
```

Comment:
Can we add an assert for memory allocation failure? I think we need to handle memory allocation errors instead of ignoring them. It can be done in a follow-up patch.

Reply:
Memory allocation failures need to be handled.  This is an error that klocworks will report and must be handled when integrated libfabric into software products.